lib/server.js

```diff
@@ -41,123 +207,59 @@ function createServer (options, httpHandler) {
-      }
-
-      const cb = typeof args[args.length - 1] === 'function' ? args.pop() : undefined
-      const options = { cb: cb }
-
-      const firstArg = args[0]
-      const argsLength = args.length
-      const lastArg = args[argsLength - 1]
-      /* Deal with listen (options) || (handle[, backlog]) */
-      if (typeof firstArg === 'object' && firstArg !== null) {
-        options.backlog = argsLength > 1 ? lastArg : undefined
-        Object.assign(options, firstArg)
-      } else if (typeof firstArg === 'string' && isNaN(firstArg)) {
-        /* Deal with listen (pipe[, backlog]) */
-        options.path = firstArg
-        options.backlog = argsLength > 1 ? lastArg : undefined
-      } else {
-        /* Deal with listen ([port[, host[, backlog]]]) */
-        options.port = argsLength >= 1 && firstArg ? firstArg : 0
-        // This will listen to what localhost is.
-        // It can be 127.0.0.1 or ::1, depending on the operating system.
-        // Fixes https://github.com/fastify/fastify/issues/1022.
-        options.host = argsLength >= 2 && args[1] ? args[1] : 'localhost'
-        options.backlog = argsLength >= 3 ? args[2] : undefined
-      }
-
-      return options
-    }
-
-    const listenOptions = normalizeListenArgs(Array.from(arguments))
-    const cb = listenOptions.cb
-
-    const wrap = err => {
-      server.removeListener('error', wrap)
-      if (!err) {
-        const address = logServerAddress()
-        cb(null, address)
-      } else {
-        this[kState].listening = false
-        cb(err, null)
-      }
-    }
+function normalizeListenArgs (args) {
+  if (args.length === 0) {
+    return { port: 0, host: 'localhost' }
+  }
 
-    const listenPromise = (listenOptions) => {
-      if (this[kState].listening && this[kState].closing) {
-        return Promise.reject(new FST_ERR_REOPENED_CLOSE_SERVER())
-      } else if (this[kState].listening) {
-        return Promise.reject(new FST_ERR_REOPENED_SERVER())
-      }
+  const cb = typeof args[args.length - 1] === 'function' ? args.pop() : undefined
+  const options = { cb: cb }
 
-      return this.ready().then(() => {
-        let errEventHandler
-        const errEvent = new Promise((resolve, reject) => {
-          errEventHandler = (err) => {
-            this[kState].listening = false
-            reject(err)
-          }
-          server.once('error', errEventHandler)
-        })
-        const listen = new Promise((resolve, reject) => {
-          server.listen(listenOptions, () => {
-            server.removeListener('error', errEventHandler)
-            resolve(logServerAddress())
-          })
-          // we set it afterwards because listen can throw
-          this[kState].listening = true
-        })
+  const firstArg = args[0]
+  const argsLength = args.length
+  const lastArg = args[argsLength - 1]
+  /* Deal with listen (options) || (handle[, backlog]) */
+  if (typeof firstArg === 'object' && firstArg !== null) {
```

Comment:
I think we should introduce a breaking change where `.listen` only accepts an object. Start with a deprecation notice for v4, but remove the variadic signature for v5.

Reply:
Could you open an issue to discuss it?

I think it would be bad to lose the `app.listen(8080)` shortcut